templatefitter/fit_model/fit_object_managers.py

```diff
@@ -0,0 +1,360 @@
+
+        # Check order and consistency of fraction parameters
+        for channel in self._channels:
+            par_i = 0
+            comps_and_temps = [(c, t) for c in channel.components for t in c.sub_templates]
+
+            assert len(comps_and_temps) == channel.total_number_of_templates, (
+                len(comps_and_temps),
+                channel.total_number_of_templates,
+            )
+            assert len(channel.fractions_mask) == channel.total_number_of_templates, (
+                len(channel.fractions_mask),
+                channel.total_number_of_templates,
+            )
+
+            last_mask_value = False
+            for counter, ((component, template), mask_value) in enumerate(zip(comps_and_temps, channel.fractions_mask)):
+                if mask_value:
+                    assert component.has_fractions
+                    assert fraction_parameter_infos[par_i].parameter_type == ParameterHandler.fraction_parameter_type, (
+                        par_i,
+                        fraction_parameter_infos[par_i].as_string(),
+                    )
+                    temp_serial_num = fraction_model_parameters[par_i].usage_serial_number_list[channel.channel_index]
+                    assert temp_serial_num == template.serial_number, (temp_serial_num, template.serial_number)
+                    temp_param = fraction_model_parameters[par_i].usage_template_parameter_list[channel.channel_index]
+                    assert template.fraction_parameter == temp_param, (
+                        template.fraction_parameter.as_string() if template.fraction_parameter else None,
+                        temp_param.as_string(),
+                    )
+
+                elif (not mask_value) and last_mask_value:
+                    assert component.has_fractions
+                    assert component.template_serial_numbers[-1] == template.serial_number, (
+                        component.template_serial_numbers[-1],
+                        template.serial_number,
+                    )
+                    assert template.fraction_parameter is None
+                    par_i += 1
+                    assert par_i <= len(fraction_parameter_infos), (par_i, len(fraction_parameter_infos))
+                else:
+                    assert (not mask_value) and (not last_mask_value), (mask_value, last_mask_value)
+                    assert not component.has_fractions
+                    assert template.fraction_parameter is None
+
+                if counter == len(comps_and_temps) - 1:
+                    assert par_i == len(fraction_parameter_infos), (counter, par_i, len(fraction_parameter_infos))
+
+                last_mask_value = mask_value
+
+        return True
+
+
+FitObject = TypeVar("FitObject", Template, Component)
+
+
+class FitObjectManager(MutableMapping[Union[str, int], FitObject]):
+    _allowed_types: Tuple[Type, ...] = (Template, Component)
```

Comment:
Wouldn't it make sense to replace this with `FitObject.__constraints__` to avoid this duplication or even replace the checks inside the methods with this? I think the checks this is needed for would also be performed by a static checker but if runtime checking is necessary this could ensure consistency.

Reply:
Thanks! Somehow I didn't really find a lot of documentation for this feature of TypeVar but it seems to work.

Reply:
Unfortunately the `FitObject.__constraints__` variable is not valid / should not be used outside of the type context, so this solution is not an option. However, the other part of your comment is also true. The static type check is in place and we also do check if the types of all objects stored in the container are the same, now. So I will probably remove this check to avoid the duplicates.

Reply:
We can always come back to the simple solution I added here, if we feel it's necessary.